core/src/main/antlr4/com/databricks/labs/remorph/parsers/tsql/TSqlParser.g4

```diff
@@ -2789,14 +2789,21 @@ predicate
     ;
 
 queryExpression
+    : unionExpression sqlIntersection*
```

Comment:
I suspect this and the below can be greatly simplified using something such as
```
queryExpression
 : querySpecification
 | queryExpression UNION ALL queryExpression
 | queryExpression INTERSECT queryExpression
etc...
```

Reply:
@ericvergnaud: This is interesting. I don't fully understand why the grammar is as it is. I already came up with this which also seems to be correct and is closer to what I would think of as the 'classical' way to express these rules:
```antlr
queryExpression
    : LPAREN queryExpression RPAREN
    | queryExpression (UNION ALL? | EXCEPT) queryExpression
    | queryExpression INTERSECT queryExpression
    | querySpecification
    ;
```

Are there performance/implementation reasons for preferring the existing folding (??) style over this?

Reply:
This is probably correct, precedence is generally best done as one rule with precedence then being explicit, then label the alts and the visitor is more simple as well

Reply:
Thanks @jimidle, @vil1 indeed improved it as follows:
```antlr
queryExpression
    : LPAREN queryExpression RPAREN                                         #inParen
    | left = queryExpression (UNION ALL? | EXCEPT) right = queryExpression  #union
    | left = queryExpression INTERSECT right = queryExpression              #intersect
    | querySpecification                                                    #simple
    ;
```

Reply:
Generally, don' tuse the labels left/right as they generate more code. ctx.queryExpression(0), ctx.queryExpression(1) is more efficient. See the read me about working with antlr grammar, labels are useful for disambiguation, but even then it might mean splitting rules or labeling the alts.

Reply:
> @ericvergnaud: This is interesting. I don't fully understand why the grammar is as it is. I already came up with this which also seems to be correct and is closer to what I would think of as the 'classical' way to express these rules:
> 
> ```antlr
> queryExpression
>     : LPAREN queryExpression RPAREN
>     | queryExpression (UNION ALL? | EXCEPT) queryExpression
>     | queryExpression INTERSECT queryExpression
>     | querySpecification
>     ;
> ```
> 
> Are there performance/implementation reasons for preferring the existing folding (??) style over this?

BTW the grammar is like this because it was just typed in by reading the MS docs, which don't encapsulate such things. You should have seen it 6 months ago ;). These are the occasions where we get to fix it up.

Reply:
Refactored to use this.

Reply:
@ericvergnaud: could you check this? Merging is blocked currently because you requested this be changed, and it has been.